op-service: Harden Transaction Manager

trianglesphere opened the pull request:
**Description**
This commit fixes a common source of errors in the transaction manager which then cause significant problems for the batcher. The transaction manager is designed to be resilient from errors when resubmitting txns or when polling for transaction receipts. It was not designed to be resilient on the initial transaction submission, and that was the root cause of several incidents.

This commit fixes this issue by wrapping `craftTx` in a retry. If there is a sustained problem for longer than the amount of retries, issues could still happen, but this will significantly reduce the number of issues.

A failure in craftTx is so harmful is because the transaction manager is wrapped in a txmgr.Queue which handles multiple in flight transactions. The queue uses an errgroup to manage concurrency and when any single txmgr.Send fails it will cancel the context & cancel the rest of the in flight sends. Because txmgr.Send could fail when creating a transaction, a transient failure would cancel multiple in flight transactions. Some of these in flight transactions would eventually land on L1 and the batcher would lose track of which frames it had submitted & thus could submit duplicate frames.

Two examples of this flow are provided in the logs. First a timeout happens in the transaction creation, then multiple transactions are cancelled via the context. Then there is a log for "aborted transaction sending". This log occurs because a transaction that was cancelled landed on L1 and the nonce of the transaction is too low. That then cancels then pending transactions again.

```
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=104,419
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="aborted transaction sending"         data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=860
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="failed to create the tx: eth_signTransaction failed: Post \"--snip--\": context deadline exceeded" data_size=120,000

t=2023-07-13T16:07:37+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=104,670
t=2023-07-13T16:07:37+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:07:37+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:07:37+0000 lvl=warn msg="unable to publish tx"                   err="aborted transaction sending"         data_size=120,000
t=2023-07-13T16:05:05+0000 lvl=warn msg="unable to publish tx"                  err="context canceled" data_size=120,000
t=2023-07-13T16:05:05+0000 lvl=warn msg="unable to publish tx"                  err="context canceled" data_size=110,635
t=2023-07-13T16:05:05+0000 lvl=warn msg="unable to publish tx"                  err="context canceled" data_size=120,000
t=2023-07-13T16:05:05+0000 lvl=warn msg="unable to publish tx"                  err="context canceled" data_size=120,000
t=2023-07-13T16:05:05+0000 lvl=warn msg="unable to publish tx"                  err="failed to create the tx: failed to get gas price info: failed to fetch the suggested gas tip cap: Post \"--snip--\": context deadline exceeded" data_size=120,000
```

**Tests**

One test change for the following reason. It also depends on #6805 in order
to make sure that the context cancellation is always returned.

The context in txmgr.Send is now being checked for cancellations by the
retry package which wraps craftTx. This caused an error in one of the
txmgr queue tests. The flow of the test is that it would send a valid tx,
send an invalid tx, and then send another tx. It'd expect that the last
tx got cancelled because of the context cancellation propogation (i.e. if
one tx in the queue fails, all will fail).

Because the context is cancelled prior to the call to txmgr.Send & this
check on the context, the transaction is no longer failing in the receipt
loop, but at craftTx. This test checked that the transaction would make it
to the receipt loop (via hooking into sendTx where it recorded the nonce).

**Metadata**

- Fixes CLI-4305

Inphi commented:
@Mergifyio update

Inphi commented:
@trianglesphere this needs a rebase. I merged develop to apply fix #6729 and unblock your PR, but now the op-service-tests fails due to other changes in develop.

tynes commented:
Rerunning CI as it may have been a flake

trianglesphere commented:
> Rerunning CI as it may have been a flake

I can reproduce the flake locally. I need to spend some time looking into it.

trianglesphere committed:
op-service: Harden Transaction Manager

This commit fixes a common source of errors in the transaction manager
which then cause significant problems for the batcher. The transaction
manager is designed to be resilient from errors when resubmitting txns or
when polling for transaction receipts. It was not designed to be resilient
on the initial transaction submission, and that was the root cause of
several incidents.

This commit fixes this issue by wrapping `craftTx` in a retry. If there is a
sustained problem for longer than the amount of retries, issues could still
happen, but this will significantly reduce the number of issues.

A failure in craftTx is so harmful is because the transaction manager is
wrapped in a txmgr.Queue which handles multiple in flight transactions.
The queue uses an errgroup to manage concurrency and when any single
txmgr.Send fails it will cancel the context & cancel the rest of the in flight
sends. Because txmgr.Send could fail when creating a transaction, a transient
failure would cancel multiple in flight transactions. Some of these in flight
transactions would eventually land on L1 and the batcher would lose track of
which frames it had submitted & thus could submit duplicate frames.

Two examples of this flow are provided in the logs. First a timeout happens in
the transaction creation, then multiple transactions are cancelled via the context.
Then there is a log for "aborted transaction sending". This log occurs because a
transaction that was cancelled landed on L1 and the nonce of the transaction is too
low. That then cancels then pending transactions again.

t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=104,419
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:22:28+0000 lvl=warn msg="unable to publish tx"                   err="aborted transaction sending"         data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=860
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:20:24+0000 lvl=warn msg="unable to publish tx"                   err="failed to create the tx: eth_signTransaction failed: Post \"--snip--\": context deadline exceeded" data_size=120,000

t=2023-07-13T16:07:37+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=104,670
t=2023-07-13T16:07:37+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:07:37+0000 lvl=warn msg="unable to publish tx"                   err="context canceled"                    data_size=120,000
t=2023-07-13T16:07:37+0000 lvl=warn msg="unable to publish tx"                   err="aborted transaction sending"         data_size=120,000
t=2023-07-13T16:05:05+0000 lvl=warn msg="unable to publish tx"                  err="context canceled" data_size=120,000
t=2023-07-13T16:05:05+0000 lvl=warn msg="unable to publish tx"                  err="context canceled" data_size=110,635
t=2023-07-13T16:05:05+0000 lvl=warn msg="unable to publish tx"                  err="context canceled" data_size=120,000
t=2023-07-13T16:05:05+0000 lvl=warn msg="unable to publish tx"                  err="context canceled" data_size=120,000
t=2023-07-13T16:05:05+0000 lvl=warn msg="unable to publish tx"                  err="failed to create the tx: failed to get gas price info: failed to fetch the suggested gas tip cap: Post \"--snip--\": context deadline exceeded" data_size=120,000

trianglesphere committed:
op-service: Fix txmgr test

The context in txmgr.Send is now being checked for cancellations by the
retry package which wraps craftTx. This caused an error in one of the
txmgr queue tests. The flow of the test is that it would send a valid tx,
send an invalid tx, and then send another tx. It'd expect that the last
tx got cancelled because of the context cancellation propogation (i.e. if
one tx in the queue fails, all will fail).

Because the context is cancelled prior to the call to txmgr.Send & this
check on the context, the transaction is no longer failing in the receipt
loop, but at craftTx. This test checked that the transaction would make it
to the receipt loop (via hooking into sendTx where it recorded the nonce).

mergify[bot] committed:
Merge branch 'develop' into jg/txmgr_harden
